csharp/ql/src/semmle/code/csharp/exprs/Expr.qll

```diff
@@ -283,19 +282,148 @@ class IsTypeExpr extends IsExpr {
   TypeAccess getTypeAccess() { result = typeAccess }
 }
 
-/**
- * An `is` pattern expression, for example `x is string s`.
- */
-class IsPatternExpr extends IsTypeExpr {
+/** An `is` pattern expression, for example `x is string s`. */
+class IsPatternExpr extends IsExpr {
   LocalVariableDeclExpr typeDecl;
 
-  IsPatternExpr() { typeDecl = this.getChild(2) }
+  IsPatternExpr() { typeDecl = this.getChild(2) and not this instanceof IsRecursivePatternExpr }
 
   /**
    * Gets the local variable declaration in this `is` pattern expression.
    * For example `string s` in `x is string s`.
    */
   LocalVariableDeclExpr getVariableDeclExpr() { result = typeDecl }
+
+  /**
+   * Gets the type being accessed in this `is` expression, for example `string`
+   * in `x is string`.
+   */
+  Type getCheckedType() { result = getTypeAccess().getTarget() }
+
+  /**
+   * Gets the type access in this `is` expression, for example `string` in
+   * `x is string`.
+   */
+  TypeAccess getTypeAccess() { result = this.getChild(1) }
+}
+
+/** An `is` recursive pattern expression, for example `x is string { Length: 5 } s`. */
+class IsRecursivePatternExpr extends IsExpr {
+  RecursivePatternExpr recursivePattern;
+
+  IsRecursivePatternExpr() { recursivePattern = this.getChild(4) }
+
+  override string toString() { result = "... is { ... }" }
+
+  /**
+   * Gets the local variable declaration in this `is` pattern expression, if any.
+   * For example `s` in `x is string { Length: 0 } s`.
+   */
+  LocalVariableDeclExpr getVariableDeclExpr() { result = this.getChild(2) }
+
+  /**
+   * Gets the recursive pattern of this `is` pattern expression. For example,
+   * `{ Length: 5 }` in `x is string { Length: 5 } s`.
+   */
+  RecursivePatternExpr getRecursivePattern() { result = recursivePattern }
+}
+
+/**
+ * A recursive pattern expression, for example `string { Length: 5 } s`
+ */
+class RecursivePatternExpr extends Expr, @recursive_pattern_expr {
```

Comment:
Why do we need a new DB type for this, couldn't it be defined instead as a union `@recursive_pattern_expr = @property_pattern_expr  | @positional_pattern_expr`?

Reply:
They're not mutually exclusive.